charms/reactive/altrelations.py

```diff
@@ -41,9 +49,574 @@ def from_name(cls, relation_name):
+    :data:`~charms.reactive.helpers.context` namespace. Handlers can access the
+    endpoint objects from there, even if that endpoint doesn't have a joined
+    relation. The name of the endpoint object is based on the endpoints
+    configured in `metadata.yaml`. Note that endpoints will only be created
+    **if the interface layer of that endpoint is included in the `layer.yaml`
+    file.**
+
+    Endpoint handlers can iterate over the list of joined relations for an
+    endpoint via the :meth:`~charms.reactive.altrelations.Endpoint.relations`
+    collection.
+    """
+
+    @classmethod
+    def from_name(cls, relation_name):
+        return getattr(context.endpoints, relation_name, None)
+
+    @classmethod
+    def from_flag(cls, flag):
+        if '.' not in flag:
+            return None
+        parts = flag.split('.')
+        if parts[0] == 'endpoint':
+            return cls.from_name(parts[1])
+        else:
+            # some older handlers might not use the 'endpoint' prefix
+            return cls.from_name(parts[0])
+
+    @classmethod
+    def _startup(cls):
+        """
+        Populate context and manage automatic relation flags.
+        """
+        for relation_name in sorted(hookenv.relation_types()):
+            # populate context based on attached relations
+            relf = relation_factory(relation_name)
+            if not issubclass(relf, cls):
+                continue
+
+            rids = sorted(hookenv.relation_ids(relation_name))
+            endpoint = relf(relation_name, rids)
+            setattr(context.endpoints, relation_name, endpoint)
+            endpoint._manage_flags()
+            for relation in endpoint.relations:
+                hookenv.atexit(relation._flush_data)
+
+    def __init__(self, relation_name, relation_ids=None):
+        self._relation_name = relation_name
+        self._relations = KeyList(map(Relation, relation_ids or []),
+                                  key='relation_id')
+        self._all_units = None
+
+    @property
+    def relation_name(self):
+        """
+        Relation name of this endpoint.
+        """
+        return self._relation_name
+
+    @property
+    def relations(self):
+        """
+        Collection of `Relation`s that are established for this `Endpoint`.
+
+        This is a `KeyList`, so it can be iterated and indexed as a list,
+        or you can look up relations by their ID.  For example::
+
+            rel0 = endpoint.relations[0]
+            assert rel0 is endpoint.relations[rel0.relation_id]
+            assert all(rel is endpoint.relations[rel.relation_id]
+                       for rel in endpoint.relations)
+            print(', '.join(endpoint.relations.keys()))
+        """
+        return self._relations
+
+    @property
+    def joined(self):
+        """
+        Whether this endpoint has remote applications attached to it.
+        """
+        return len(self.relations) > 0
+
+    def flag(self, flag):
```

Comment:
This was really more of a helper for populating the `{endpoint_name}` (I'm inclined to change that to match the method name, discussed above) portion in a consistent way, and to provide a shortcut if you didn't want to type that bit out all the time (probably a bad idea?).  I'm not sure how it would be used to document public flags for the layer since they happen at runtime; can you explain your thoughts there more?

I had also toyed around with the idea of having some class-level structure to pre-define the flags, but you still couldn't populate the `{endpoint_name}` portion until runtime, so I'm not sure how we could make that work.  But that would certainly give us a great way to be able to add doc strings for the flags themselves, perhaps even integrated into the definition itself.